### PR TITLE
test(benchmarks): improve bechmarks to have complete performance data

### LINKED ISSUE #311: docs: add section to help choose merkle tree
There are several Merkle Tree "flavors:
- incremental
- sparse
- leant
- persistent
- ...

It would be helpful to write a comparison of of all them, answering questions like
- "when not to use"/"when to use"
- "used by?": protocol xyz, app xyz...
- include benchmarks: maybe one merkle tree is faster at generating a proof, but slower to update?

in order to help developers make a choice depending on their use case.

### PR DESCRIPTION
**Pull request by prodasl**
## Description

Added the missing Sparse Merkle tree, the missing operations:

- Added Sparse Merkle Tree to the benchmark suites to complete the available merkle tree implementations in zk-kit.

- Added operation to the benchmark suites (proof generation, proof verification, update, delete

- Changed the "imt.ts" name to "benchmark-merkle-trees.ts"

- Added README.md documentation about the benchmark suite

- Added log managment with winston library to handle any error produced during the benchmark withtout break the benchmark execution

## Related Issue(s)
Closes #311

## Other information

The results folder is present in gitignore then the results are not included in the pull request, but you can use the README.md instructions to generate it.

## Checklist
It would be helpful to write a comparison of of all them, answering questions like
-   [ ] "when not to use"/"when to use"
-   [ ]  "used by?": protocol xyz, app xyz...
-   [x] include benchmarks: maybe one merkle tree is faster at generating a proof, but slower to update?

### CODE REVIEW

**Comment by prodasl** [comment]
Dear @sripwoud, I have made the changes and run the recommended yarn commands before commit, but I have received the error: 
"YN0028: The lockfile would have been modified by this install, which is explicitly forbidden."

**Please give me some clues about how to solve it**, I used **node 20.17.0 and yarn 4.1.0**, when I run yarn to install the dependencies the yarn.lock has been changed but I didn't include it to the commit because I thought that is not convenient.

**Comment by prodasl** [comment]
Hello, I have already fixed the problems in validation of the PR.

File: benchmarks/benchmark-merkle-trees.ts [hunk 1]
```diff
@@ -0,0 +1,340 @@
...
+            }
+        }),
+        b.add(`SparseMT - Generated ${samples} proofs`, () => {
+            try {
+                if (leafSMT < samples) {
+                    sparseMerkleTree.createProof(sample[leafSMT].toString(16))
+                } else {
+                    leafSMT = 0
+                }
+                leafSMT += 1
```
Comment (sripwoud) [inline benchmarks/benchmark-merkle-trees.ts:163]: I am not sure I am understand how the leaf index/count is supposed behave.
If leaf count === samples, leaf count will be set to 1 (first 0 then increased by 1). Is it what we want?

File: benchmarks/README.md [hunk 2]
```diff
@@ -0,0 +1,100 @@
...
+Fastest is IMT - Add 1000 leaves
+
+Suite: proof-generation-merkle-trees-1000
+IMT - Generated 500 proofs x 4,321 ops/sec ±2.02% (80 runs sampled)
+LeanIMT - Generated 500 proofs x 3,456 ops/sec ±1.75% (78 runs sampled)
+SparseMT - Generated 500 proofs x 2,789 ops/sec ±2.63% (68 runs sampled)
+Fastest is IMT - Generated 500 proofs
+```
+
+The Benchmarks suggested in the index.ts are for **8, 128 and 2014 leafs** to see how each Merkle tree for different sizes of trees
```
Comment (sripwoud) [inline benchmarks/README.md:83]:
```suggestion
The Benchmarks suggested in the index.ts are for **8, 128 and 1024 leafs** to see how each Merkle tree for different sizes of trees
```

File: benchmarks/index.ts [hunk 3]
```diff
@@ -1,13 +1,16 @@
...
-}
+// Benchmark with 128 leafs
+treeDepth = 7
+numberOfLeaves = 2 ** treeDepth
+run(treeDepth, numberOfLeaves)
+
+// Benchmark with 1024 leafs
+treeDepth = 10
+numberOfLeaves = 2 ** treeDepth
+run(treeDepth, numberOfLeaves)
```
Comment (sripwoud) [inline benchmarks/index.ts:16]: refactor with a loop?

File: benchmarks/benchmark-merkle-trees.ts [hunk 4]
```diff
@@ -0,0 +1,340 @@
...
+            logger.error(error)
+        }
+    }
+
+    /*  Benchmarking Proof Verification:
+            This section benchmarks proof verification for each tree type. */
+    leafIMT = 0
+    leafLeanIMT = 0
+    leafSMT = 0
+    name = `proof-verfication-merkle-trees-${numberOfLeaves}`
```
Comment (sripwoud) [inline benchmarks/benchmark-merkle-trees.ts:200]:
```suggestion
    name = `proof-verification-merkle-trees-${numberOfLeaves}`
```

**Review by sripwoud — CHANGES_REQUESTED** [review]
Nice!
Small requests for changes.

it would be nice to add a summary results table to the root README (and update the conclusion if necessary based on the analysis).

Eg https://gist.github.com/sripwoud/06a1602fe8dd41c72f39c317cd1676f5#file-summary-md

File: benchmarks/benchmark-merkle-trees.ts [hunk 5]
```diff
@@ -0,0 +1,340 @@
+/* This script benchmarks three implementations of Merkle Trees—Incremental Merkle Tree,
+Lean Incremental Merkle Tree, and Sparse Merkle Tree—to compare their performance across
+various operations such as adding leaves, generating proofs, and verifying proofs.
+The results are saved as .html files (charts and tables) for analysis. */
+import b from "benny"
```
[the remark this answers is not in the document]
Comment (prodasl) [reply benchmarks/benchmark-merkle-trees.ts:5]: It had already added before I started working in the repository. Talking about that I had problems when I tried to add to-hex and winston repositories as development dependency in the monorepo's root, **Github gave an error when it was validating the PR** I asked in comments to have a clue about how should I resolve it, but then I used another way to transform to HEX and put some instructions in the benchmarks/README.md as workaround.

File: benchmarks/benchmark-merkle-trees.ts [hunk 6]
```diff
@@ -0,0 +1,340 @@
...
+import { poseidon2 } from "poseidon-lite"
+import sha256 from "crypto-js/sha256"
+import winston from "winston"
```
[the remark this answers is not in the document]
Comment (prodasl) [reply benchmarks/benchmark-merkle-trees.ts:8]: I used the workaround I mentioned in the last comment.

File: benchmarks/README.md [hunk 7]
```diff
@@ -0,0 +1,100 @@
...
+## Installation
+
+Before running the benchmark, ensure that you have Node.js version 20 installed and the required dependencies set up.
+
+1. **Clone the Repository**:
+    ```bash
+    git clone https://github.com/privacy-scaling-explorations/zk-kit.git
+    cd zk-kit
+    yarn
+    yarn winston
```
[the remark this answers is not in the document]
Comment (prodasl) [reply benchmarks/README.md:30]: This is the mentioned workaround

File: benchmarks/benchmark-merkle-trees.ts (hunk 1 above)
Comment (prodasl) [reply benchmarks/benchmark-merkle-trees.ts:163]: **Benny does not have a setting to limit the number of cycles** will be performed then I put this code to restart doing the operations from the first leaf, giving the user the possibility to use the same benchmark to test trees of any sizes without unnecessary errors (index overflow) during the benchmark execution.

File: benchmarks/benchmark-merkle-trees.ts [hunk 8]
```diff
@@ -0,0 +1,340 @@
...
+            }
+        }
+
+        return sample
+    }
+
+    const incrementalMerkleTree = new IMT(poseidon2, numberOfLeaves, 0, 2)
+    const leanIncrementalMerkleTree = new LeanIMT((a, b) => poseidon2([a, b]))
+    const hash = (childNodes: ChildNodes) => sha256(childNodes.join("")).toString()
+    const sparseMerkleTree = new SMT(hash)
```
[the remark this answers is not in the document]
Comment (prodasl) [reply benchmarks/benchmark-merkle-trees.ts:53]: This change can break the logic, we need the hash instantiated before

File: benchmarks/benchmark-merkle-trees.ts (hunk 5 above)
[the remark this answers is not in the document]
Comment (prodasl) [reply benchmarks/benchmark-merkle-trees.ts:5]: It had been already added before I started working on the issue, then this is why you don't see it in the changes track

File: benchmarks/benchmark-merkle-trees.ts (hunk 6 above)
[the remark this answers is not in the document]
Comment (prodasl) [reply benchmarks/benchmark-merkle-trees.ts:8]: I have added it in the last commint today

File: benchmarks/index.ts (hunk 3 above)
Comment (prodasl) [reply benchmarks/index.ts]: I have made the refactoring in the last commit of today

File: benchmarks/benchmark-merkle-trees.ts (hunk 1 above)
Comment (prodasl) [reply benchmarks/benchmark-merkle-trees.ts]: You right, I have changed it in the last commit

**Comment by prodasl** [comment]
I have already took action in all the requested changes, please @sripwoud  let me know how it looks

File: benchmarks/benchmark-merkle-trees.ts (hunk 8 above)
[the remark this answers is not in the document]
Comment (sripwoud) [reply benchmarks/benchmark-merkle-trees.ts:53]: ups yes your absolutely right

**Comment by sripwoud** [comment]
@prodasl 
> YN0028: The lockfile would have been modified by this install, which is explicitly forbidden.

is this solved, or do you still need help with this?

File: yarn.lock
Comment (sripwoud) [inline yarn.lock:1]: the indentation of this file is broken leading to an error upon running `yarn`:
```
yarn
➤ YN0088: A new stable version of Yarn is available: 4.5.0!
➤ YN0088: Upgrade now by running yarn set version 4.5.0

YAMLException: bad indentation of a mapping entry at line 3174, column 3:
      languageName: unknown
      ^
      languageName: unknown
```

Was it necessary to edit `yarn.lock` manually? Cases where one has to do so are very rare.

File: benchmarks/README.md [hunk 9]
```diff
@@ -0,0 +1,99 @@
...
+    cd zk-kit
+    yarn
+    ```
+
+## Usage
+
+To run the benchmark, use the following command:
+
+```bash
+npx ts-node benchmarks/index.ts
```
Comment (sripwoud) [inline benchmarks/README.md:37]: can't we mention the already defined `benchmarks` script here instead?
https://github.com/privacy-scaling-explorations/zk-kit/blob/2173bc800d2a22409678a7d3983df48d4f8efd18/package.json#L24

**Review by sripwoud — CHANGES_REQUESTED** [review]
nACK

One cannot install the deps with `yarn` or run yarn scripts (like `yarn benchmarks`) due to the lockfile's broken indentation.

**Comment by prodasl** [comment]
> @prodasl
> 
> > YN0028: The lockfile would have been modified by this install, which is explicitly forbidden.
> 
> is this solved, or do you still need help with this?

**Comment by prodasl** [comment]
> @prodasl
> 
> > YN0028: The lockfile would have been modified by this install, which is explicitly forbidden.
> 
> is this solved, or do you still need help with this?

Hi, @sripwoud, thanks for the support. 
I have some doubts about it yet, it is related to my try to send the changes on yarn.lock, because the first time I did the PR, I haven't sent the changes of yarn.lock, but this time I have a lot of change reported in the yarn lock related to packages versions, the only real change was the winston lib added, then my questions are:

**Should I include the changes of yarn.lock or definitely it should not be part of the PR? (when I not included I got the mentioned problem YN0028)**
**If I include the yarn.lock, should I include the package version changes or just the new line for winston lib?**

File: benchmarks/README.md (hunk 9 above)
Comment (prodasl) [reply benchmarks/README.md]: Thanks, I have already changed it.

File: yarn.lock
Comment (prodasl) [reply yarn.lock:1]: Yes, I tried to sent the added winston line, instead of all the packages version changes that "yarn" command made in my local version of yarn.lock, but It broken the indentation, as i mentioned in the another comment about (YN0028), I need guidance about how handle the yarn.lock in the PR and commits flow.

**Comment by prodasl** [comment]
Finally, I tried sharing the yarn.lock as it is in my local environment, sorry for all the trouble but I have not worked with those files before, all the checks in Github looks like are passed and all the change requests covered.

Please let me know if it is ready. (thanks again)

**Comment by sripwoud** [comment]
> > @prodasl
> 
> > 
> 
> > > YN0028: The lockfile would have been modified by this install, which is explicitly forbidden.
> 
> > 
> 
> > is this solved, or do you still need help with this?
> 
> 
> 
> Hi, @sripwoud, thanks for the support. 
> 
> I have some doubts about it yet, it is related to my try to send the changes on yarn.lock, because the first time I did the PR, I haven't sent the changes of yarn.lock, but this time I have a lot of change reported in the yarn lock related to packages versions, the only real change was the winston lib added, then my questions are:
> 
> 
> 
> **Should I include the changes of yarn.lock or definitely it should not be part of the PR? (when I not included I got the mentioned problem YN0028)**
> 
> **If I include the yarn.lock, should I include the package version changes or just the new line for winston lib?**
> 
> 
> 
> 
> 
> 

TLDR: you should commit any changes to yarn.lock every time you change deps in package.json

If there are more changes in yarn.lock than strictly related to your PR it is okay and normal: it is due to yarn bumping deps automatically according to semantic versioning rules.

In practice devs exceptionally review changes in yarn.lock because there are always so huge. Reviewing package.json changes is enough.

But avoid modifying yarn.lock yourself. Stick to the yarn add/remove commands.

If it happens that there a mismatch between the committed yarn.lock and package.json then you get the error:

YN0028: The lockfile would have been modified by this install, which is explicitly forbidden.

File: benchmarks/README.md [hunk 10]
```diff
@@ -0,0 +1,99 @@
...
+
+The benchmark compares the performance of each Merkle Tree implementation for operations like adding leaves, generating proofs, and verifying proofs.
+
+## Table of Contents
+
+-   [Installation](#installation)
+-   [Usage](#usage)
+-   [Benchmark](#benchmark)
+-   [Results](#results)
+-   [Dependencies](#dependencies)
```
Comment (sripwoud) [inline benchmarks/README.md:17]: nit:
nice but unnecessary.
GitHub web ui automatically adds a button to see the table of contents for all `.md` files :wink: 
https://github.blog/changelog/2021-04-13-table-of-contents-support-in-markdown-files/

image

**Review by sripwoud — APPROVED** [review]
tACK (successfully ran `yarn benchmarks` locally)

Thanks @prodasl ! Nice job.
